File: concepts/04_collections_and_iteration/01_iteration_protocol/r/test_02_iterable_iterator_and_fallbacks.R
# polyglot-family: collections_and_iteration
# polyglot-concept: iteration_protocol
# polyglot-related: languages/r/language/08_collection_workflows_and_tabular_data/
# polyglot-related+: test_059_apply_lapply_sapply_and_vapply_contracts.R
#
# 共同问题：iterable 与 iterator 是否分离；缺少协议时采用什么 fallback。
# 对照观察：R 不暴露 next() 式 iterator 对象；索引序列、`lapply` 和 closure 是惯用替代。

make_next <- local({
    index <- 0L
    values <- c("a", "b")
    function() {
        index <<- index + 1L
        if (index > length(values)) return(NULL)
        values[[index]]
    }
})

stopifnot(
    identical(make_next(), "a"),
    identical(make_next(), "b"),
    is.null(make_next())
)

File: concepts/04_collections_and_iteration/01_iteration_protocol/lua/test_02_iterable_iterator_and_fallbacks.lua
-- Common question: how are custom iterables, iterator state, and fallbacks represented?
-- Inputs: a closure iterator, generic-for triple, __pairs, and a plain table.
-- Observations: explicit iterator/state/control values, closure state, and metamethod fallback.
-- polyglot-family: collections_and_iteration
-- polyglot-concept: iteration_protocol
-- polyglot-related: languages/lua/language/06_metatables_and_objects/
-- polyglot-related+: test_044_operator_and_iteration_metamethods.lua

local t = require("support.assertions")

local function iterator(state, control)
    local next_index = control + 1
    local value = state[next_index]
    if value ~= nil then return next_index, value end
end

local collected = {}
for index, value in iterator, {"a", "b"}, 0 do
    collected[index] = value
end
t.equal(table.concat(collected), "ab")

local wrapped = setmetatable({storage = {answer = 42}}, {
    __pairs = function(self) return next, self.storage, nil end,
})
local key, value = next(wrapped.storage)
t.equal(key, "answer")
t.equal(value, 42)
for observed_key, observed_value in pairs(wrapped) do
    t.equal(observed_key, "answer")
    t.equal(observed_value, 42)
end

t.done()
